File: apps/mobile/ios/Happy/Models/Friend.swift
//
//  Friend.swift
//  Happy
//
//  Copyright (c) 2024-2026 Enflame Media. All rights reserved.
//

import Foundation

/// Represents a friend connection between two Happy users.
///
/// Friends can see each other's online status and share Claude Code sessions
/// in real-time. The friend model mirrors the server-side representation
/// and supports Codable for JSON serialization.
struct Friend: Identifiable, Codable, Hashable {
    /// Unique identifier for the friend relationship.
    let id: String

    /// The friend's user/account ID.
    let userId: String

    /// The friend's display name.
    var displayName: String

    /// The friend's email address, if available.
    var email: String?

    /// The friend's avatar URL, if available.
    var avatarUrl: String?

    /// The friend's current online status.
    var status: FriendStatus

    /// When the friendship was established.
    let friendsSince: Date

    /// The last time this friend was seen online (nil if never or currently online).
    var lastSeenAt: Date?

    /// Number of shared sessions with this friend.
    var sharedSessionCount: Int

    /// Whether this friend is currently online.
    var isOnline: Bool {
        status == .online
    }
}

// MARK: - Friend Status

/// The online status of a friend.
enum FriendStatus: String, Codable, Hashable {
    /// The friend is currently online and active.
    case online

    /// The friend is online but idle/away.
    case away

    /// The friend is currently in a Claude Code session.
    case inSession = "in_session"

    /// The friend is offline.
    case offline

    /// Display text for the status.
    var displayText: String {
        switch self {
        case .online: return "Online"
        case .away: return "Away"
        case .inSession: return "In Session"
        case .offline: return "Offline"
        }
    }

    /// System image name for the status indicator.
    var systemImageName: String {
        switch self {
        case .online: return "circle.fill"
        case .away: return "moon.fill"
        case .inSession: return "terminal.fill"
        case .offline: return "circle"
        }
    }
}

// MARK: - Sample Data

extension Friend {
    static let sample = Friend(
        id: "friend-001",
        userId: "user-123",
        displayName: "Alice Developer",
        email: "alice@example.com",
        avatarUrl: nil,
        status: .online,
        friendsSince: Date().addingTimeInterval(-86400 * 30),
        lastSeenAt: nil,
        sharedSessionCount: 5
    )

    static let sampleOffline = Friend(
        id: "friend-002",
        userId: "user-456",
        displayName: "Bob Engineer",
        email: "bob@example.com",
        avatarUrl: nil,
        status: .offline,
        friendsSince: Date().addingTimeInterval(-86400 * 60),
        lastSeenAt: Date().addingTimeInterval(-3600),
        sharedSessionCount: 2
    )

    static let sampleInSession = Friend(
        id: "friend-003",
        userId: "user-789",
        displayName: "Carol Coder",
        email: nil,
        avatarUrl: nil,
        status: .inSession,
        friendsSince: Date().addingTimeInterval(-86400 * 10),
        lastSeenAt: nil,
        sharedSessionCount: 0
    )

    static let samples: [Friend] = [sample, sampleOffline, sampleInSession]
}

File: apps/mobile/android/app/src/main/java/com/enflame/happy/domain/model/Friend.kt
package com.enflame.happy.domain.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Domain model representing a friend connection between two Happy users.
 *
 * Friends can see each other's online status and share Claude Code sessions
 * in real-time. This model mirrors the server-side representation.
 *
 * @property id Unique identifier for the friend relationship.
 * @property displayName The friend's display name.
 * @property username The friend's username for identification.
 * @property avatarUrl The friend's avatar URL, if available.
 * @property status The friend's current online status.
 * @property lastSeen Timestamp of when the friend was last seen online (null if currently online).
 * @property addedAt Timestamp of when the friendship was established.
 * @property sharedSessionCount Number of shared sessions with this friend.
 */
@Serializable
data class Friend(
    val id: String,
    val displayName: String,
    val username: String,
    val avatarUrl: String? = null,
    val status: FriendStatus = FriendStatus.OFFLINE,
    val lastSeen: Long? = null,
    val addedAt: Long,
    val sharedSessionCount: Int = 0
) {
    /**
     * Whether this friend is currently online (any non-offline status).
     */
    val isOnline: Boolean
        get() = status != FriendStatus.OFFLINE
}

/**
 * Online status of a friend.
 */
@Serializable
enum class FriendStatus {
    /** The friend is currently online and active. */
    @SerialName("online")
    ONLINE,

    /** The friend is online but idle/away. */
    @SerialName("away")
    AWAY,

    /** The friend is currently in a Claude Code session. */
    @SerialName("in_session")
    IN_SESSION,

    /** The friend is offline. */
    @SerialName("offline")
    OFFLINE
}

/**
 * Domain model representing a friend request.
 *
 * @property id Unique identifier for the friend request.
 * @property fromUser The user who sent the request.
 * @property toUser The user who received the request.
 * @property status The current status of the request.
 * @property message Optional message included with the request.
 * @property createdAt Timestamp of when the request was created.
 */
@Serializable
data class FriendRequest(
    val id: String,
    val fromUser: FriendRequestUser,
    val toUser: FriendRequestUser,
    val status: FriendRequestStatus = FriendRequestStatus.PENDING,
    val message: String? = null,
    val createdAt: Long
)

/**
 * Minimal user info included in a friend request.
 *
 * @property id The user's unique identifier.
 * @property displayName The user's display name.
 * @property username The user's username.
 * @property avatarUrl The user's avatar URL, if available.
 */
@Serializable
data class FriendRequestUser(
    val id: String,
    val displayName: String,
    val username: String,
    val avatarUrl: String? = null
)

/**
 * Status of a friend request.
 */
@Serializable
enum class FriendRequestStatus {
    /** The request is awaiting a response. */
    @SerialName("pending")
    PENDING,

    /** The request has been accepted. */
    @SerialName("accepted")
    ACCEPTED,

    /** The request has been declined. */
    @SerialName("declined")
    DECLINED
}
